src/stan/services/pathfinder/multi.hpp

```diff
@@ -105,136 +109,214 @@ inline int pathfinder_lbfgs_multi(
     ParamWriter& parameter_writer, DiagnosticWriter& diagnostic_writer,
     bool calculate_lp = true, bool psis_resample = true) {
   const auto start_pathfinders_time = std::chrono::steady_clock::now();
-  std::vector<std::string> param_names;
-  param_names.push_back("lp_approx__");
-  param_names.push_back("lp__");
-  model.constrained_param_names(param_names, true, true);
-
-  parameter_writer(param_names);
-  std::vector<Eigen::Array<double, Eigen::Dynamic, 1>> individual_lp_ratios;
-  individual_lp_ratios.resize(num_paths);
-  std::vector<Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic>>
-      individual_samples;
-  individual_samples.resize(num_paths);
   std::atomic<size_t> lp_calls{0};
+  // Save idx of pathfinder and it's elbo for resampling later
+  tbb::concurrent_vector<std::pair<Eigen::Index, internal::elbo_est_t>>
+      elbo_estimates;
+  elbo_estimates.reserve(num_paths + 10);
+  auto constrain_fun = [](auto&& constrained_draws, auto&& unconstrained_draws,
+                          auto&& model, auto&& rng) {
+    model.write_array(rng, unconstrained_draws, constrained_draws);
+    return constrained_draws;
+  };
   try {
+    stan::callbacks::concurrent_writer safe_write{parameter_writer};
     tbb::parallel_for(
         tbb::blocked_range<int>(0, num_paths), [&](tbb::blocked_range<int> r) {
           for (int iter = r.begin(); iter < r.end(); ++iter) {
-            auto pathfinder_ret
-                = stan::services::pathfinder::pathfinder_lbfgs_single<true>(
-                    model, *(init[iter]), random_seed, stride_id + iter,
-                    init_radius, history_size, init_alpha, tol_obj, tol_rel_obj,
-                    tol_grad, tol_rel_grad, tol_param, num_iterations,
-                    num_elbo_draws, num_draws, save_iterations, refresh,
-                    interrupt, logger, init_writers[iter],
-                    single_path_parameter_writer[iter],
-                    single_path_diagnostic_writer[iter], calculate_lp);
-            if (unlikely(std::get<0>(pathfinder_ret) != error_codes::OK)) {
-              logger.error(std::string("Pathfinder iteration: ")
-                           + std::to_string(iter) + " failed.");
-              return;
+            if (psis_resample && calculate_lp) {
```

Comment:
Thoughts on hoisting this outside the parallel-for and having two separate ones? It probably doesn't matter for speed, but I think it could help readability